File: Java/Prototype/ShapeFactory/src/com/ericgtkb/ShapeFactory.java
package com.ericgtkb;

import java.util.HashMap;
import java.util.Map;

public class ShapeFactory {
    private static Map<String, Shape> shapePrototypes = new HashMap<>();

    public static Shape getShape(String shapeType) throws CloneNotSupportedException {
        Shape shape = shapePrototypes.get(shapeType);
        return (Shape) shape.clone();
    }

    public static void setUpPrototypes() {
        Rectangle rectangle = new Rectangle();
        Triangle triangle = new Triangle();

        shapePrototypes.put("rectangle", rectangle);
        shapePrototypes.put("triangle", triangle);
    }
}

File: C++/Prototype/ShapeFactory/shape_factory.cc
#include "shape_factory.h"


std::unordered_map<std::string, std::unique_ptr<Shape>> ShapeFactory::shape_prototypes {};

std::unique_ptr<Shape> ShapeFactory::get_shape(const std::string & shape_type) {
    Shape* shape {shape_prototypes[shape_type].get()};
    return shape->clone();
}

void ShapeFactory::set_up_prototypes() {
    shape_prototypes["rectangle"] = std::make_unique<Rectangle>();
    shape_prototypes["triangle"] = std::make_unique<Triangle>();
}
